gprofiler/profilers/java.py

```diff
@@ -277,12 +272,46 @@ def get_java_version(process: Process, stop_event: Event) -> str:
     # 2. assume JDK type by the path, e.g the "java" Docker image has
     #    "/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java" which means "OpenJDK". needs to be checked for
     #    other JDK types.
-    java_path = f"/proc/{nspid}/exe"
+    java_path: Optional[str] = None
+    if is_java_basename(process):
+        nspid = get_process_nspid(process.pid)
+        java_path = f"/proc/{nspid}/exe"  # it's a symlink and will be resolveable under process' mnt ns
+    else:
+        libjvm_path: Optional[str] = None
+        for m in process.memory_maps():
+            if re.match(DETECTED_JAVA_PROCESSES_REGEX, m.path):
+                libjvm_path = m.path
+                break
+        if libjvm_path is None:
+            return None
+        libjvm_dir = os.path.dirname(libjvm_path)
+        # support two java layouts - it's either lib/server/../../bin/java or lib/{arch}/server/../../../bin/java:
+        java_candidate_paths = [
+            Path(libjvm_dir, "../../bin/java").resolve(),
+            Path(libjvm_dir, "../../../bin/java").resolve(),
+        ]
+        for p in java_candidate_paths:
+            # don't need resolve_proc_root_links here - paths in /proc/pid/maps are normalized.
+            proc_relative_path = Path(f"/proc/{process.pid}/root", p.relative_to("/"))
+            if proc_relative_path.exists():
+                if os.access(proc_relative_path, os.X_OK):
+                    java_path = str(p)
+                    break
+    return java_path
+
+
+# process is hashable and the same process instance compares equal
+@functools.lru_cache(maxsize=_JAVA_VERSION_CACHE_MAX)
+def get_java_version(process: Process, stop_event: Event) -> Optional[str]:
+    # make sure we're able to find "java" binary bundled with process libjvm
+    process_java_path = _get_process_ns_java_path(process)
+    if process_java_path is None:
+        return None
 
     def _run_java_version() -> "CompletedProcess[bytes]":
         return run_process(
             [
-                java_path,
+                process_java_path,  # type: ignore # checked to be not-None above
```

Comment:
```suggestion
                cast(str, process_java_path),
```
always nicer than `type: ignore`

Reply:
👍